[GEOT-5846] Upgrade JDom to version 2.0.6

mprins opened the pull request:
This upgrades the JDom dependency, 1.1.3 to 2.0.6 and resolves [GEOT-5846](https://osgeo-org.atlassian.net/browse/GEOT-5846)

Some related work may be necessary, both here and in eg. Geoserver and there are some deprecated JDom methods in use that could be refactored as well. Also the old jdom library may still be dragged in by other dependencies such as JTS.

The upgrade procedure essentially consists of:
- updating the Maven artifact 
- and renaming the packages.
- fixing the incompatibilites

There are some minor incompatibilities in the JDom api, notably the  ` XPath#selectNodes(....)` method which returns a `java.util.List<>` with a different capture (`<?>` instead of `<Element>`)

jodygarnett commented:
This look good, any progress to report?

mprins commented:
This is ready for review; I summarized the changes below. 

### Files that have an API change

`...app-schema/app-schema/src/main/java/org/geotools/data/complex/xml/XmlResponse.java`
`...app-schema/app-schema/src/main/java/org/geotools/data/complex/xml/XmlXpathFilterData.java`
`...app-schema/app-schema/src/main/java/org/geotools/util/XmlXpathUtilites.java`

Because these now have and org.jdom2.Document instead of org.jdom.Document (different package) as an accessible class variable or public method argument. 


`...modules/library/coverage/src/main/java/org/geotools/coverage/grid/io/imageio/geotiff/GeoTiffIIOMetadataEncoder.java`
`...modules/unsupported/geotiff_new/src/main/java/org/geotools/coverage/grid/imageio/geotiff/metadata/GeoTiffIIOMetadataEncoder.java`

Because it is now using org.jdom2.Element instead of org.jdom.Element as a method argument 


### Files that have a change in the throws of a public method

`...modules/library/main/src/main/java/org/geotools/data/ows/ServiceExceptionParser.java`
`...coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CreateIndexer.java`
`...plugin/shapefile/src/main/java/org/geotools/data/shapefile/shp/xml/ShpXmlFileReader.java`

Now throw org.jdom2.JDOMException instead of org.jdom.JDOMException

### Files that don't have a public API change

`...modules/library/main/src/main/java/org/geotools/data/ows/Response.java`
`...modules/plugin/geotiff/src/main/java/org/geotools/gce/geotiff/GeoTiffWriter.java`
`...unsupported/app-schema/webservice/src/main/java/org/geotools/data/ws/WS_DataStore.java`
`...unsupported/app-schema/webservice/src/main/java/org/geotools/filter/AsXpathFunctionExpression.java`
`...unsupported/geotiff_new/src/main/java/org/geotools/gce/geotiff/GeoTiffUtils.java`
`...unsupported/app-schema/webservice/src/test/java/org/geotools/data/complex/XmlDataStoreTest.java` (NB test code)

aaime commented:
Think we need @bencaradocdavies to have a look (app-schema is involved).

aaime commented on modules/plugin/coverage-multidim/netcdf/src/test/java/org/geotools/coverage/io/netcdf/CreateIndexer.java:
```diff
@@ -279,27 +281,29 @@ private static void addDomainsToCoverage(String schemaAttribs, StringBuilder bui
-                    if (!elevationAttributes.contains(name)) {
-                        elevationAttributes.add(name);
+        for (Object e : schemaAttributes) {
+        	if (e instanceof Element) {
```
Does not follow coding conventions (since this part was fully intented anyways,  you could select it and IDE reformat it)

aaime commented:
So, I guess GeoServer build will break solid once this one is merged? If so, we cannot merge until we have also the GeoServer follow up pull request ready on the "merge" button :-)

mprins commented:
hmmm, fixed the formatting and rebased to master and now see a test failure where before all passed: `JDBCTransactionStateTest.testSetTransactionNullWithInternalConnection:100->testLogWarnings:118 Regular close, no warnings expected. expected:<0> but was:<1>`

I've started on a geoserver PR as well

bencaradocdavies commented:
Close-reopen to retrigger Travis CI with the fix for [GEOT-5869](https://osgeo-org.atlassian.net/browse/GEOT-5869).

bencaradocdavies commented:
Travis CI now passes. The app-schema changes look straightforward. I will try a local build to see if GeoServer app-schema online tests are affected.

bencaradocdavies commented:
Full local build + GWC +  GeoServer with https://github.com/geoserver/geoserver/pull/2636 + GeoServer app-schema online tests (postgis) + GeoServer community build (`-DskipTests`) all pass for me.

Are there objections to merging this, and then https://github.com/geoserver/geoserver/pull/2636 when GeoTools master Jenkins has passed?

aaime commented:
@bencaradocdavies none here, if you feel like following the progress of the two builds, go for it! :-)

bencaradocdavies commented:
Sigh. A spurious JDOM 1 import introduced in https://github.com/geotools/geotools/pull/1735 three hours ago breaks the JDOM 2 upgrade I just merged. Fix is trivial.

bencaradocdavies commented:
Fixed on master:

Remove spurious JDOM 1 import introduced for [GEOS-8227] (#1735) that breaks the build after JDOM 2 upgrade [GEOT-5846] (#1719):
https://github.com/geotools/geotools/commit/4139108b6bdba84902ca82be4082a1b6a23805ca
